blog/_posts/2016-10-06-julia-0.5-highlights.md

```diff
@@ -0,0 +1,726 @@
+     (-0.1,0.4)  (-0.1,0.3)  (-0.1,0.2)  (-0.1,0.1)
+
+Except that with the vectorized call syntax, the matrix of pairs never has to be constructed, nor does the messy code to apply `clip` to each element of `B` together with the corresponding `lo` and `hi` need to be written.
+When `B` is larger than it is in this toy example, avoiding constructing a matrix of `(lo,hi)` pairs can be a big efficiency win.
+
+There is a bit more to the vectorized calls syntax story.
+It's common to write expressions applying multiple vectorized functions to some arrays.
+For example, one might write something like:
+
+    atan2(cos(X), sin(Y))
+
+Currently, this allocates two temporary intermediate arrays – one to store `cos(X)` and one for `sin(Y)`.
+If we use vectorized function call syntax, however, these calls are syntactically fused into a *single* call to `broadcast` with an anonymous function.
+In other words, `atan2.(cos.(X), sin.(Y))` becomes
+
+    broadcast((x, y)->atan2(cos(x), sin(y)), X, Y)
+
+This computation avoids allocating any intermediate arrays and performs the entire vectorized computation in one fell swoop into the result array.
+We can see this difference in allocation when we benchmark these expressions:
+
+    julia> using BenchmarkTools
+
+    julia> X, Y = rand(1000,1000), rand(1000,1000);
+
+    julia> @benchmark atan2(cos(X), sin(Y))
+    BenchmarkTools.Trial:
+      memory estimate:  22.89 mb
+      allocs estimate:  13
+      minimum time:     54.09 ms (1.32% GC)
+      median time:      55.82 ms (4.30% GC)
+      mean time:        55.89 ms (4.25% GC)
+      maximum time:     56.95 ms (4.36% GC)
+
+    julia> @benchmark atan2.(cos.(X), sin.(Y))
+    BenchmarkTools.Trial:
+      memory estimate:  7.63 mb
+      allocs estimate:  38
+      minimum time:     51.83 ms (0.00% GC)
+      median time:      52.18 ms (0.00% GC)
+      mean time:        52.94 ms (1.48% GC)
+      maximum time:     57.91 ms (4.17% GC)
+
+    julia> 22.89/7.63
+    3.0
+
+The speed gain is marginal, but allocations are reduced by a factor of 3 as predicted – only the output array is allocated, no temporaries are.
+But the real reduction in allocation comes when we use the new `.=` operator to also do vectorized assignment:
+
+    julia> Z = zeros(X);
+
+    julia> @benchmark Z .= atan2.(cos.(X), sin.(Y))
+    BenchmarkTools.Trial:
+      memory estimate:  96.00 bytes
+      allocs estimate:  3
+      minimum time:     50.76 ms (0.00% GC)
+      median time:      50.98 ms (0.00% GC)
+      mean time:        51.08 ms (0.00% GC)
+      maximum time:     52.23 ms (0.00% GC)
+
+With in-place vectorized assignment, we can fill the pre-allocated array, `Z`, without doing any allocation (the 96 bytes is an artifact).
+The fact that this *still* isn't that much faster than the version that allocates several temporary and an output array is primarily a testament to how fast and carefully optimized Julia's allocator is;
```

Comment:
An example for which fusion provides impressive speedup:
```julia
julia> @benchmark max.(abs.($A), abs.($B))
BenchmarkTools.Trial:
  samples:          1298
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  7.63 mb
  allocs estimate:  12
  minimum time:     2.71 ms (0.00% GC)
  median time:      3.20 ms (0.00% GC)
  mean time:        3.84 ms (20.32% GC)
  maximum time:     6.95 ms (42.08% GC)
```
on master, versus
```julia
julia> @benchmark max(abs($A), abs($B))
BenchmarkTools.Trial:
  samples:          301
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  22.89 mb
  allocs estimate:  8
  minimum time:     14.98 ms (15.85% GC)
  median time:      16.41 ms (15.26% GC)
  mean time:        16.63 ms (15.39% GC)
  maximum time:     20.15 ms (13.21% GC)
```
on 0.4. The `atan2`/`sin`/`cos` example is charming though. Best!